### PR TITLE
Adding LUPID for intercomm create and merge

### PR DESCRIPTION
**Pull request by charlesarcher**
Rebased and migrated from : https://github.com/wesbland/mpich-mirror/pull/25

Will continue the review here.  @yfguo @raffenet

### CODE REVIEW

**Comment by charlesarcher** [comment]
@yguo:  I get a compiler error:

```In file included from /home/cjarcher/code/mpich/tmp/tmp.U7H0CTxpdV/ssg_sfi-mpich/src/mpid/ch4/include/mpidch4.h:252:0,
                 from /home/cjarcher/code/mpich/tmp/tmp.U7H0CTxpdV/ssg_sfi-mpich/src/mpid/ch4/include/mpidpost.h:15,
                 from /home/cjarcher/code/mpich/tmp/tmp.U7H0CTxpdV/ssg_sfi-mpich/src/include/mpiimpl.h:234,
                 from /home/cjarcher/code/mpich/tmp/tmp.U7H0CTxpdV/ssg_sfi-mpich/src/mpi/coll/barrier_group.c:8:
/home/cjarcher/code/mpich/tmp/tmp.U7H0CTxpdV/ssg_sfi-mpich/src/mpid/ch4/netmod/include/netmod_impl.h:141:26: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before \
‘int’
 MPI_STATIC_INLINE_PREFIX int MPIDI_NM_upids_to_lupids(int size, size_t * remote_upid_size,```

**Comment by charlesarcher** [comment]
Need to convert MPI-->MPL in the inline prefix.  I can fix this.

**Comment by charlesarcher** [comment]
Someone from ANL should review this too.  There are a lot of upper level changes.

File: src/mpid/ch4/netmod/ucx/ucx_init.h [hunk 1]
```diff
@@ -353,56 +353,63 @@ static inline int MPIDI_NMI_allocate_address_table()
...
+    int i, total_size = 0;
+    char *temp_buf = NULL, *curr_ptr = NULL;
+    char keyS[MPIDI_UCX_KVSAPPSTRLEN];
+    char valS[MPIDI_UCX_KVSAPPSTRLEN];
     int len = MPIDI_UCX_global.max_addr_len;
 
-    MPIDIU_comm_rank_to_pid(comm_ptr, rank, &lpid, &avtid);
-    MPIR_Assert(rank < comm_ptr->local_size);
+    *local_upid_size = (size_t*) MPL_malloc(comm->local_size * sizeof(size_t));
+    temp_buf = (char*) MPL_malloc(comm->local_size * len);
```
Comment (charlesarcher) [inline src/mpid/ch4/netmod/ucx/ucx_init.h]: MPIR_CHKLMEM_ here?

**Comment by charlesarcher** [comment]
Some more warnings:
```
/home/cjarcher/code/mpich/tmp/tmp.oB84DY6fZg/ssg_sfi-mpich/src/mpid/ch4/src/ch4_init.h:665:5: warning: ‘remote_size’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/cjarcher/code/mpich/tmp/tmp.oB84DY6fZg/ssg_sfi-mpich/src/mpid/ch4/src/ch4_init.h:624:75: warning: ‘remote_node_ids’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/cjarcher/code/mpich/tmp/tmp.oB84DY6fZg/ssg_sfi-mpich/src/mpid/ch4/include/mpidpre.h:439:78: warning: ‘lpid’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/cjarcher/code/mpich/tmp/tmp.oB84DY6fZg/ssg_sfi-mpich/src/mpid/ch4/src/ch4_init.h:665:5: warning: ‘remote_size’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/cjarcher/code/mpich/tmp/tmp.oB84DY6fZg/ssg_sfi-mpich/src/mpid/ch4/src/ch4_init.h:624:75: warning: ‘remote_node_ids’ may be used uninitialized in this function [-Wmaybe-uninitialized]
/home/cjarcher/code/mpich/tmp/tmp.oB84DY6fZg/ssg_sfi-mpich/src/mpid/ch4/include/mpidpre.h:439:78: warning: ‘lpid’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```

**Comment by charlesarcher** [comment]
I reviewed this patch.  If we can address the warnings, I think it can go in.  I'd like to test it a bit more, too.

@yfguo @raffenet any more comments from ANL?

File: src/mpid/ch3/include/mpidpost.h [hunk 2]
```diff
@@ -179,6 +179,10 @@ int MPID_GPID_GetAllInComm( MPIR_Comm *comm_ptr, int local_size,
 			    MPIR_Gpid local_gpids[], int *singlePG );
 int MPID_GPID_Get( MPIR_Comm *comm_ptr, int rank, MPIR_Gpid *gpid );
 int MPID_GPID_ToLpidArray( int size, MPIR_Gpid gpid[], int lpid[] );
+int MPID_intercomm_exchange_map(MPIR_Comm *local_comm_ptr, int local_leader,
```
Comment (raffenet) [inline src/mpid/ch3/include/mpidpost.h]: nit: capitalize Intercomm

File: src/mpid/ch3/src/mpid_vc.c [hunk 3]
```diff
@@ -424,6 +424,225 @@ int MPID_GPID_ToLpidArray( int size, MPIR_Gpid in_gpid[], int lpid[] )
...
+#define FCNAME MPL_QUOTE(FUNCNAME)
+static inline int MPIDI_LPID_GetAllInComm(MPIR_Comm *comm_ptr, int local_size,
+                                          int local_lpids[])
+{
+    int i;
+    MPIR_Assert( comm_ptr->local_size == local_size );
+    for (i=0; i<comm_ptr->local_size; i++) {
+	(void)MPID_Comm_get_lpid( comm_ptr, i, &local_lpids[i], FALSE );
+    }
+    return 0;
```
Comment (raffenet) [inline src/mpid/ch3/src/mpid_vc.c]: Just make the function void instead of always returning zero? That or actually check return values of `MPID_Comm_get_lpid` and exit on error.

File: src/mpid/ch4/src/ch4i_comm.h [hunk 4]
```diff
@@ -1144,4 +1144,68 @@ MPL_STATIC_INLINE_PREFIX int MPIDII_comm_create_rank_map(MPIR_Comm * comm)
     return mpi_errno;
 }
 
+#undef FUNCNAME
+#define FUNCNAME MPIDII_check_disjoint_lupids
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4i_comm.h]: `MPIDII` prefix should be `MPIDI`. Ditto for any other instances in the series.

File: src/mpid/ch3/src/mpid_vc.c [hunk 5]
```diff
@@ -549,7 +549,7 @@ int MPID_intercomm_exchange_map(MPIR_Comm *local_comm_ptr, int local_leader,
                                       peer_comm_ptr, MPI_STATUS_IGNORE, &errflag );
         if (mpi_errno) MPIR_ERR_POP(mpi_errno);
 
-        MPL_DBG_MSG_FMT(MPIID_CH3_DBG_OTHER,VERBOSE,(MPL_DBG_FDEST, "local size = %d, remote size = %d", local_size,
+        MPL_DBG_MSG_FMT(MPIDI_CH3_DBG_OTHER,VERBOSE,(MPL_DBG_FDEST, "local size = %d, remote size = %d", local_size,
```
Comment (raffenet) [inline src/mpid/ch3/src/mpid_vc.c]: extract this typo fix to separate commit.

File: src/mpid/ch3/src/mpid_vc.c [hunk 6]
```diff
@@ -331,7 +331,7 @@ int MPID_GPID_GetAllInComm( MPIR_Comm *comm_ptr, int local_size,
 	}
 	*gpid++ = vc->pg_rank;
 
-        MPL_DBG_MSG_FMT(MPIR_DBG_COMM,VERBOSE, (MPL_DBG_FDEST,
+        MPL_DBG_MSG_FMT(MPIDI_CH3_DBG_OTHER,VERBOSE, (MPL_DBG_FDEST,
```
Comment (raffenet) [inline src/mpid/ch3/src/mpid_vc.c]: Probably put this in separate commit, as it's not related to the functionality changes.

File: src/mpid/ch4/netmod/ofi/ofi_init.h [hunk 7]
```diff
@@ -631,62 +630,76 @@ static inline int MPIDI_NM_comm_get_lpid(MPIR_Comm * comm_ptr,
...
-    MPIR_Assert(rank < comm_ptr->local_size);
-    size_t sz = sizeof(MPIDI_OFI_GPID(gpid).addr);
-    MPIDI_OFI_CALL(fi_av_lookup(MPIDI_Global.av, MPIDI_OFI_COMM_TO_PHYS(comm_ptr, rank),
-                                &MPIDI_OFI_GPID(gpid).addr, &sz), avlookup);
-    MPIR_Assert(sz <= sizeof(MPIDI_OFI_GPID(gpid).addr));
+    int i, total_size = 0;
+    char *temp_buf = NULL, *curr_ptr = NULL;
+
+    *local_upid_size = (size_t*) MPL_malloc(comm->local_size * sizeof(size_t));
+    temp_buf = (char*) MPL_malloc(comm->local_size * MPIDI_Global.addrnamelen);
```
Comment (yfguo) [inline src/mpid/ch4/netmod/ofi/ofi_init.h]: Do we have a coding convention about this? @raffenet

File: src/mpid/ch4/netmod/ofi/ofi_init.h [hunk 8]
```diff
@@ -649,6 +649,38 @@ static inline int MPIDI_NM_gpid_get(MPIR_Comm * comm_ptr, int rank, MPIR_Gpid *
...
+
+    for (i = 0; i < comm->local_size; i++) {
+        (*local_upid_size)[i] = MPIDI_Global.addrnamelen;
+        MPIDI_OFI_CALL(fi_av_lookup(MPIDI_Global.av, MPIDI_OFI_COMM_TO_PHYS(comm, i),
+                                            &temp_buf[i * MPIDI_Global.addrnamelen],
+                                            &(*local_upid_size)[i]), avlookup);
+        total_size += (*local_upid_size)[i];
```
Comment (yfguo) [inline src/mpid/ch4/netmod/ofi/ofi_init.h]: Yes. This is use to handle the mixed address sizes. I can adjust the ADI a little bit so that OFI can skip this if it knows the address size is fixed.

File: src/mpid/ch3/src/mpid_vc.c (hunk 3 above)
Comment (yfguo) [inline src/mpid/ch3/src/mpid_vc.c]: I think actually checking the return of `MPID_Comm_get_lpid` is better. Something like `ret |= MPID_Comm_get_lpid( ... );` and `return ret;` would be sufficient. Note that `MPID_Comm_get_lpid()` always do `return MPI_SUCCESS`, so the compiler should eventually optimize it. I think `void` function is not the best choice because we will have to change the function definition again if we later decide to add error code here. Unused return value can be optimized by compiler.

Comment (raffenet) [inline src/mpid/ch3/src/mpid_vc.c]: K, sounds good to me.

**Comment by charlesarcher** [comment]
Superceded by #2381, closing.
